### PR TITLE
xbps-src: add sccache support

### PR DESCRIPTION
**Pull request by Logarithmus**
This PR adds support for using `sccache` when building `cargo` crates with `xbps-src`, just like `ccache`.
Unfortunately, this PR is blocked by https://github.com/void-linux/void-packages/pull/29756.

Roadmap:
- [ ] Perform some benchmarks of build time speedups and share the results.
- [ ] Test caching crates across different packages.
- [ ] Implement `rust_bootstrap` flag into `xbps-src`, see #30702 (comment)
- [ ] Test parallel builds in different masterdirs, see https://github.com/void-linux/void-packages/pull/30702#issuecomment-833552762

### CODE REVIEW

File: srcpkgs/base-chroot/template [hunk 1]
```diff
@@ -17,6 +17,6 @@ esac
 depends+="
  base-files binutils gcc gcc-ada libada-devel
  patch sed findutils diffutils make gzip coreutils
- file bsdtar ccache xbps mpfr ncurses libreadline8
+ file bsdtar ccache rust-sccache>=0.2.15 xbps mpfr ncurses libreadline8
```
Comment (q66) [inline srcpkgs/base-chroot/template:20]: nope

**Review by q66 — CHANGES_REQUESTED** [review]

File: srcpkgs/base-chroot/template (hunk 1 above)
Comment (Logarithmus) [reply srcpkgs/base-chroot/template:20]: Why?

Comment (q66) [reply srcpkgs/base-chroot/template:20]: base-chroot only depends on bootstrap packages and sccache is not one, and will not be one, as that'd impose rust toolchain dependency for bootstrap and we!re not doing that

besides, sccache is limited to a small fraction of CPU architectures, which is another reason even if rust was bootstrap

Comment (Logarithmus) [reply srcpkgs/base-chroot/template:20]: OK, I'll remove `rust-sccache` from the dependencies. Instead, I'll add a note into README.md that in order to use `sccache` with `xbps-src`, one should run `xbps-install -r masterdir/ rust-sccache`.

Comment (Logarithmus) [reply srcpkgs/base-chroot/template:20]: > besides, sccache is limited to a small fraction of CPU architectures, which is another reason even if rust was bootstrap

AFAIK plain `sccache` (local cache) should work as long as you have Rust for the architecture. `sccache-dist` (distributed compilation) works on Linux x86_64 only, but this PR isn't about `sccache-dist`.

**Comment by Piraty** [comment]
interesting. i looked into icecc for xbps-src myself, scache could be a viable alternative.

File: srcpkgs/base-chroot/template (hunk 1 above)
Comment (q66) [reply srcpkgs/base-chroot/template:20]: it doesn't, read the template you are updating

Comment (q66) [reply srcpkgs/base-chroot/template:20]: (but also we're not gonna make rust toolchain a bootstrap package, so the point is moot)

Comment (q66) [reply srcpkgs/base-chroot/template]: also, installing manually into masterdir won't work either

if you want to do this you will need to find a way to have the masterdir lazily install it the first time you use it, similarly to the crosstoolchain and so on

Comment (ericonr) [reply srcpkgs/base-chroot/template]: Have the cargo build style pull it in if we aren't building `rust`, `cargo` or `sccache`? With a `rust_bootstrap` flag or whatever.

File: common/build-helper/rust.sh [hunk 2]
```diff
@@ -4,6 +4,10 @@
 export CARGO_BUILD_JOBS="$XBPS_MAKEJOBS"
 export CARGO_HOME="/host/cargo"
 
+if [ "$XBPS_SCCACHE" = "yes" ]; then
```
Comment (Duncaen) [inline common/build-helper/rust.sh:7]: I don't think we ever test for "yes", this should be `[ -n "$XBPS_SCCACHE" ]`.

File: srcpkgs/base-chroot/template (hunk 1 above)
Comment (Logarithmus) [reply srcpkgs/base-chroot/template]: > also, installing manually into masterdir won't work either
> 
> if you want to do this you will need to find a way to have the masterdir lazily install it the first time you use it, similarly to the crosstoolchain and so on

I've just checked it out: `./xbps-src clean` doesn't remove manually installed `sccache` from the `masterdir/`. `hostdir/sccache` is also preserved. Then if I try to build some Rust package (e. g. `fd`), `SCCACHE_DIR=hostdir/binpkgs sccache -s` reports cache hits.

https://user-images.githubusercontent.com/29541480/117307385-3356a100-ae89-11eb-9f77-44f8c6b50de1.mp4

File: common/build-helper/rust.sh (hunk 2 above)
Comment (Logarithmus) [reply common/build-helper/rust.sh]: Fixed

**Comment by Johnnynator** [comment]
This will blow up once you run parallel builds in different masterdirs. sccache always listens on some tcp port on localhost. You would need to figure out how to either allocate a TCP port per masterdir or implement the communication via unix sockets and upstream it.

**Comment by Logarithmus** [comment]
> This will blow up once you run parallel builds in different masterdirs. sccache always listens on some tcp port on localhost. You would need to figure out how to either allocate a TCP port per masterdir or implement the communication via unix sockets and upstream it.

I should definitely test this scenario. I still think this PR can be merged in its current state, because `$XBPS_SCCACHE` is turned off by default. Maybe I should add a disclaimer in `README.md` about parallel builds in differrent masterdirs though.

**Comment by ericonr** [comment]
- how faster do builds get?
- does it work for cooperation between different packages using the same crate?
- I'm against the need to install the package manually, it should be automated in some way (I suggested a solution above)

**Comment by Logarithmus** [comment]
> * how faster do builds get?
> * does it work for cooperation between different packages using the same crate?
> * I'm against the need to install the package manually, it should be automated in some way (I suggested a solution above)
> 
> And it can't be merged if it's waiting on another PR.

 * I'll perform some benchmarks and share the results.
 * I think it caches crates across different packages, but more testing is needed to proof that.
 * OK, I'll try adding `rust_bootstrap` into `xbps-src`, just like you've suggested in https://github.com/void-linux/void-packages/pull/30702#discussion_r627392158

**Pull request closed without merging**